x-pack/filebeat/module/virustotal/livehunt/_meta/fields.yml

```diff
@@ -0,0 +1,924 @@
+        built or compiled. It can also be faked by malware creators.
+      type: date
+    - name: header
+      default_field: false
+      description: >
+        Header information of the ELF file.
+      release: beta
+      type: group
+      fields:
+        - name: class
+          description: >
+            Header class of the ELF file.
+          type: keyword
+        - name: data
+          description: >
+            Data table of the ELF header.
+          type: keyword
+        - name: machine
+          description: >
+            Machine architecture of the ELF header.
+          type: keyword
+        - name: os_abi
+          description: >
+            NEED TO ADD
+          type: keyword
+        - name: type
+          description: >
+            Header type of the ELF file.
+          type: keyword
+        - name: version
+          description: >
+            Version of the ELF header.
+          type: keyword
+        - name: abi_version
+          type: keyword
+          description: >
+            Version of the ELF Application Binary Interface (ABI).
+        - name: entrypoint
+          format: string
+          type: long
+          description: >
+            Header entrypoint of the ELF file.
+        - name: object_version
+          type: keyword
+          description: >
+            "0x1" for original ELF files.
+
+    - name: number_program_headers
+      description: >
+        Number of ELF Program Headers.
+      type: long
+    - name: number_section_headers
+      description: >
+        Number of ELF Section Headers.
+      type: long
+    - name: sections
+      default_field: false
+      description: >
+        Section information of the ELF file.
+      release: beta
+      type: group
+      fields:
+        - name: flags
+          description: >
+            ELF Section List flags.
+          type: keyword
+        - name: name
+          description: >
+            ELF Section List name.
+          type: keyword
+        - name: physical_offset
+          description: >
+            ELF Section List offset.
+          type: keyword
+        - name: section_type
+          description: >
+            ELF Section List type.
+          type: keyword
+        - name: size
+          description: >
+            ELF Section List size.
+          format: bytes
+          type: long
+        - name: virtual_address
+          description: >
+            ELF Section List virtual address.
+          format: string
+          type: long
+    - name: exports
+      description: >
+        List of exported element names and types
+      release: beta
+      type: group
+      fields:
+        - name: name
+          description: >
+            Name of exported symbol
+          type: keyword
+          default_field: false
+        - name: type
+          description: >
+            Type of exported symbol
+          type: keyword
+          default_field: false
+    - name: imports
```

Comment:
Wondering if it would make sense for this level to describe an actual linked in library and for the stuff currently nested here (i.e. `name`, `type`, etc.) to specify the symbols imported by a library. Otherwise you get symbols free of context from where they're actually being imported.

Reply:
I agree. Ideally, I'd like to see a common representation across ELF, PE, and Mach-O. Unfortunately, these formats don't work the same, especially in the way they import symbols. I think making exports and imports `nested` rather than a `group` makes sense to maintain context. Making these a nested dictionary with common fields for each binary type might be the right answer. Not all binary types will have all fields populated, but at least consistent across formats. I'll play with this.

Reply:
So, it's totally possible to try and resolve the libraries that symbols come from in ELF format, [see example](https://github.com/andrewstucki/fingerprint/blob/625f616f4551c33d5e984ab28a0cc48e793c7fad/fixtures/libxml2.so.fingerprint#L462-L473). There's actually default support for resolving the libraries in [Go's standard library](https://golang.org/pkg/debug/elf/#ImportedSymbol). I couldn't tell if the `ndjson` example that is dropped here actually does that as part of the VirusTotal service for ELF files, but if it does, it would probably make sense to scope these.

Reply:
I'd really like to have a common interface across PE, ELF, and MachO for this. [LIEF](https://lief.quarkslab.com/doc/latest/api/python/abstract.html) actually does this as an analytic framework, but VT doesn't expose this data equally across all binary types. We could implement a common fieldset for imports, and applications can populate it as they are able to.

Proposal:

`file.*.imported_symbols`:

```json
{
  "name": "my_symbol", "size": 0, "value": 0, "type": "function", "library_name": "my_library.dll"
}
```

In the case of PE, the VT data would permit populating symbol name, library name, and we can derive a type of "function". For ELF, the data provides symbol name and type (the samples I've seen), for Mach-O...  VT doesn't give us any symbols... just a list of linked libraries, which could feasibly go somewhere else as a flat list, say `file.*.linked_libraries`. 

Anything not provided by the source (VT in this case) would be omitted. Another application could feasibly populate this data with much greater detail. The library_name for ELF could be resolved as you say, but it's not coded in the binary specifically (I think).